app/forms/payment_form.rb

```diff
@@ -0,0 +1,78 @@
+class PaymentForm < ApplicationForm
+  attr_accessor :account_transaction,
+                :processor_token,
+                :coupon,
+                :user
+
+  before_save :decrease_stock
+  before_save :mark_coupon_as_used
+  after_save :update_account_transaction
+  after_save :charge!
+
+  validate :product_must_have_stock
+  validate :coupon_must_not_be_used
+
+  def initialize(args = {})
+    super(args)
+
+    @models = [payment]
```

Comment:
where are the "multiple models"? Maybe you can have `@models = [payment, product]` instead of decreasing stock in a callback, same for coupon